presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java

```diff
@@ -240,22 +241,65 @@ public static HiveStorageFormat getHiveStorageFormat(Map<String, Object> tablePr
         List<String> bucketedBy = getBucketedBy(tableProperties);
         List<SortingColumn> sortedBy = getSortedBy(tableProperties);
         int bucketCount = (Integer) tableProperties.get(BUCKET_COUNT_PROPERTY);
-        if ((bucketedBy.isEmpty()) && (bucketCount == 0)) {
-            if (!sortedBy.isEmpty()) {
-                throw new PrestoException(INVALID_TABLE_PROPERTY, format("%s may be specified only when %s is specified", SORTED_BY_PROPERTY, BUCKETED_BY_PROPERTY));
-            }
-            return Optional.empty();
+
+        List<String> clusteredBy = getClusteredBy(tableProperties);
+        List<Integer> clusterCount = getClusterCount(tableProperties);
+        List<String> distribution = getDistribution(tableProperties);
+
+        if (doBucketing(bucketedBy)) {
+            return getPropertyForBucketing(bucketedBy, sortedBy, bucketCount);
         }
-        if (bucketCount < 0) {
-            throw new PrestoException(INVALID_TABLE_PROPERTY, format("%s must be greater than zero", BUCKET_COUNT_PROPERTY));
+
+        if (doClustering(clusteredBy)) {
+            return getPropertyForClustering(bucketedBy, sortedBy, bucketCount, clusteredBy, clusterCount, distribution);
         }
-        if (bucketCount > 1_000_000) {
-            throw new PrestoException(INVALID_TABLE_PROPERTY, format("%s should be no more than 1000000", BUCKET_COUNT_PROPERTY));
+
+        if (!sortedBy.isEmpty()) {
+            throw new PrestoException(INVALID_TABLE_PROPERTY, format("%s may be specified only when %s is specified",
+                    SORTED_BY_PROPERTY, BUCKETED_BY_PROPERTY));
         }
-        if (bucketedBy.isEmpty() || bucketCount == 0) {
-            throw new PrestoException(INVALID_TABLE_PROPERTY, format("%s and %s must be specified together", BUCKETED_BY_PROPERTY, BUCKET_COUNT_PROPERTY));
+        return Optional.empty();
+    }
+
+    private static boolean doBucketing(List<String> bucketedBy)
+    {
+        return !bucketedBy.isEmpty();
+    }
+
+    private static Optional<HiveBucketProperty> getPropertyForBucketing(List<String> bucketedBy, List<SortingColumn> sortedBy, int bucketCount)
+    {
+        if (bucketCount < 0 || bucketCount > 1_000_000) {
+            throw new PrestoException(INVALID_TABLE_PROPERTY, format("%s must be within range (0, 1,000,000]", BUCKET_COUNT_PROPERTY));
+        }
+        return Optional.of(new HiveBucketProperty(
+                bucketedBy, bucketCount, sortedBy, HIVE_COMPATIBLE, Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty()));
+    }
+
+    private static boolean doClustering(List<String> clusteredBy)
+    {
+        return !clusteredBy.isEmpty();
+    }
+
+    private static Optional<HiveBucketProperty> getPropertyForClustering(
+            List<String> bucketedBy,
+            List<SortingColumn> sortedBy,
+            int bucketCount,
+            List<String> clusteredBy,
+            List<Integer> clusterCount,
+            List<String> distribution)
```

Comment:
We should check
- the clusteredBy size is the same as clusterCount
- clusterCount has all positive integers that are all 2^n. Also it should between something like 0 and 2^16
- the multiplication of all cluster numbers should be <= 2^16

Reply:
Right. Will check.